feat: Run order groups for ATEM commands

rjmunro opened the pull request:
<!--
Before you open a PR, be sure to read our Contribution guidelines:
https://sofie-automation.github.io/sofie-core//docs/for-developers/contribution-guidelines
-->

## About the Contributor
<!--
Tell us who / which organization you are representing, and how the Sofie team will be able to contact you.
Example: "This pull request is posted on behalf of the NRK."
-->
This pull request is posted on behalf of the BBC.

## Type of Contribution

This is a Feature

It's required for the following bugfix: https://github.com/Sofie-Automation/sofie-atem-state/pull/37.

## Current Behavior

Commands sent to an ATEM are batched where possible. You cannot control which batch they go in. Additionally the ATEM firmware sometimes seems to execute things out of order within a batch.. On some occasions, order is important, for example when setting up and switching on a key it's important not to switch it on until after it's input are all set. Also when switching it off, it's important not to make any other changes until after it has switched off. 

## New Behavior

You can set a runOrderGroup property on an ATEM command. This is a number. Commands will only be batched if they are in the same group. Groups will be run in numerical order. Default is 0, so you can use a positive number to run something after other commands (such as turning on a key after sending it's settings), or a negative number to ensure it happens first (such as turning off a key before changing settings).

## Testing Instructions

This will be easiest to test in conjunction with the corresponding PR in ATEM-state.

## Other Information

A minor version bump and release will be needed as this adds new API surface. It is backwards compatible as a default is provided which won't change behaviour.

VSCode was showing syntax problems for me because the build target for typescript was older than some of the library features being used, so I bumped the version to match `@sofie-automation/code-standard-preset`. I didn't upgrade the dependency on that project because that caused other issues which will need addressing.

## Status
<!--
Before you open the PR, make sure the items below are done.
If they're not, please open the PR as a Draft.
-->

- [x] PR is ready to be reviewed.
- [x] The functionality has been tested by the author.
- [x] Relevant unit tests has been added / updated.
- [x] Relevant documentation (code comments, [system documentation](https://sofie-automation.github.io/sofie-core//)) has been added / updated.

rjmunro committed:
feat: Add a runOrderGroup property to ATEM commands

Only batch operations with the same runOrderGroup, then run them lowest
first. When not specified, defaults to 0, so to run before other commands
use a negative value.

jstarpl reviewed (changes requested):
Looks good, but test suites need updating to match the new structure before this can be merged.

Julusian commented on tsconfig.build.json:
```diff
@@ -9,6 +9,7 @@
 		},
-		"types": ["node"]
+		"types": ["node"],
+		"target": "es2022", // TODO:RJM: Remove es2022 target when we upgrade @sofie-automation/code-standard-preset from v2 to v3
```
I don't think this is safe to do currently, as this library still technically supports node14 which only supports es2020

nytamin commented on src/commands/CommandBase.ts:
```diff
@@ -23,12 +23,14 @@ export abstract class DeserializedCommand<T> implements IDeserializedCommand {
 
 export interface ISerializableCommand {
 	serialize(version: ProtocolVersion): Buffer
+	runOrderGroup?: number
```
Could you add a comment detailing this property?
Something like
```ts
/** When set, ensures that commands with the lowest runOrderGroup are sent first. (Defaults to 0) */
```

Julusian replied:
Probably include something about methods being sent in separate batches, not fully following the original order of the commands

rjmunro committed:
Add explanatory JSDoc

rjmunro committed:
Fix tests

rjmunro committed:
feat: Allow usage of direct from git repo

E.g. when used as a dependency like in yarn with something like:

	"resolutions": {
		"atem-connection": "github:Sofie-automation/sofie-atem-connection#branchname"
	},

Julusian commented on package.json:
```diff
@@ -24,7 +24,8 @@
   "scripts": {
-    "postinstall": "./node_modules/.bin/husky install",
+	"prepare": "run build:main",
+    "postinstall": "husky install || true",
```
This reliably fails for me with `command not found: husky`.

It happens because `husky` is not a direct dependency (only via @sofie-automation/code-standard-preset) so yarn doesn't find it as a valid executable.

I suspect this is why you added the `|| true`, but this just means that husky will never be setup

Julusian commented on package.json:
```diff
@@ -24,7 +24,8 @@
   ],
   "scripts": {
-    "postinstall": "./node_modules/.bin/husky install",
+	"prepare": "run build:main",
```
I'm not 100% convinced by this prepare script. Largely because I don't know the full implications of when it will be run.
As far as I can tell from a quick google, `yarn` will never run it, `npm` will as part of publishing or installing from git.

So this looks like it will solve installing from git for npm users, it won't do anything for yarn users (ie us).
Not enough to argue against it, just worth noting that it is a half solution to installing from git.

However, this might 'break' the npm publishing now (or at least the desired flow). Since setting up trusted publishing, we are using `npm publish` and are intentionally doing the build a few steps earlier, before escalating priviledges to have the publish token. It looks like this will 'break' that flow by rebuilding the code as part of the publish command right where I was trying to avoid it
